### PR TITLE
Add taxonomy schema validate option to the command line tool

### PR DESCRIPTION
**Pull request by ronenkat**
Add taxonomy schema validate option to the command line tool.
Add test cases for a valid and invalid taxonomy.

### CODE REVIEW

File: pkg/taxonomy/schema_test.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package taxonomy
+
+import (
+	"testing"
+
+	. "github.com/onsi/gomega"
+)
+
+/*
```
Comment (simanadler) [inline pkg/taxonomy/schema_test.go:9]: Remove commented out code

File: cmd/m4d/main.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	cobra.OnInitialize(initConfig)
+	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+	return cmd
+}
+
+func initConfig() {
+	viper.AutomaticEnv()
+}
+
+func main() {
```
Comment (simanadler) [inline cmd/m4d/main.go:37]: Is this a tool that is meant to be run from the command line?  Who is meant to use it? Where are the user instructions?

File: cmd/m4d/taxonomy.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import (
+	"fmt"
+	"strings"
+
+	"github.com/mesh-for-data/mesh-for-data/pkg/taxonomy"
+	"github.com/spf13/cobra"
+	"helm.sh/helm/v3/cmd/helm/require"
+)
+
+func newTaxonomyCmd() *cobra.Command {
```
Comment (simanadler) [inline cmd/m4d/taxonomy.go:12]: I suggest adding documentation about what this does.

File: cmd/m4d/taxonomy.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	cmd := &cobra.Command{
+		Use:   "taxonomy",
+		Short: "tools for working with taxonomies",
+		Args:  require.NoArgs,
+	}
+	cmd.AddCommand(newTaxonomySchemaCmd())
+	return cmd
+}
+
+func newTaxonomySchemaCmd() *cobra.Command {
```
Comment (simanadler) [inline cmd/m4d/taxonomy.go:22]: I suggest adding documentation about what this does.

File: cmd/m4d/taxonomy.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	cmd := &cobra.Command{
+		Use:   "schema",
+		Short: "commands for working with taxonomy JSON schema files",
+		Args:  require.NoArgs,
+	}
+	cmd.AddCommand(newTaxonomyValidateCmd())
+	return cmd
+}
+
+func newTaxonomyValidateCmd() *cobra.Command {
```
Comment (simanadler) [inline cmd/m4d/taxonomy.go:32]: I suggest adding documentation about what this does.

File: pkg/taxonomy/bad_catalog.structs.schema.json [hunk 6]
```diff
@@ -0,0 +1,65 @@
+{
```
Comment (simanadler) [inline pkg/taxonomy/bad_catalog.structs.schema.json:1]: Is this meant to be a real schema or just used for testing the tool?  If the latter, perhaps it should be in a test data directory?

File: pkg/taxonomy/catalog.structs.schema.json [hunk 7]
```diff
@@ -0,0 +1,66 @@
+{
```
Comment (simanadler) [inline pkg/taxonomy/catalog.structs.schema.json:1]: Is this meant to be a real schema or just used for testing the tool? If the latter, perhaps it should be in a test data directory?  If it's the former then perhaps either update the description of the PR or separate into a separate PR.

File: pkg/taxonomy/catalog.values.schema.json [hunk 8]
```diff
@@ -0,0 +1,39 @@
+{
```
Comment (simanadler) [inline pkg/taxonomy/catalog.values.schema.json:1]: Is this meant to be a real schema or just used for testing the tool? If the latter, perhaps it should be in a test data directory?  If it's the former then perhaps either update the description of the PR or separate into a separate PR.

File: pkg/taxonomy/document.go [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+package taxonomy
+
+import (
+	"path/filepath"
+
+	"github.com/xeipuuv/gojsonschema"
+)
+
+// ValidateDocument validates input json data against a taxonomy file
+func ValidateDocument(taxonomyFile string, jsonData string) (*gojsonschema.Result, error) {
```
Comment (simanadler) [inline pkg/taxonomy/document.go:13]: Is this meant to be used only by the tool that validates taxonomies via the cli?  If not, I imagine runtime validation would not be done from a file.  Right?

File: pkg/taxonomy/ext.json [hunk 10]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (simanadler) [inline pkg/taxonomy/ext.json:1]: What is this and how is it related to the tool?

File: pkg/taxonomy/extref_test.go [hunk 11]
```diff
@@ -0,0 +1,17 @@
+package taxonomy
+
+import (
+	"testing"
+
+	"github.com/getkin/kin-openapi/openapi3"
+)
+
+func TestExternalRef(t *testing.T) {
```
Comment (simanadler) [inline pkg/taxonomy/extref_test.go:9]: Some comments would help here

File: pkg/taxonomy/schema_test.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+func TestValidateSchema_embedded_module(t *testing.T) {
+	taxonomy_file := "module.structs.schema.json"
+	g := NewGomegaWithT(t)
+	err := ValidateSchema("../../config/taxonomy/" + taxonomy_file)
+	g.Expect(err).NotTo(HaveOccurred())
+}
+*/
+/*
```
Comment (simanadler) [inline pkg/taxonomy/schema_test.go:17]: Remove commented out code

**Review by simanadler — COMMENTED** [review]
I think some documentation about who should use the tool and how to use the tool would be helpful

File: cmd/m4d/taxonomy.go (hunk 4 above)
Comment (ronenkat) [reply cmd/m4d/taxonomy.go:24]: done

File: cmd/m4d/taxonomy.go (hunk 5 above)
Comment (ronenkat) [reply cmd/m4d/taxonomy.go:35]: done

File: pkg/taxonomy/ext.json (hunk 10 above)
Comment (ronenkat) [reply pkg/taxonomy/ext.json]: will do a cleanup

File: pkg/taxonomy/extref_test.go (hunk 11 above)
Comment (ronenkat) [reply pkg/taxonomy/extref_test.go]: will do a cleanup

File: pkg/taxonomy/schema_test.go (hunk 1 above)
Comment (ronenkat) [reply pkg/taxonomy/schema_test.go]: done

File: pkg/taxonomy/schema_test.go (hunk 12 above)
Comment (ronenkat) [reply pkg/taxonomy/schema_test.go]: done

**Review by ronenkat — COMMENTED** [review]
Thank you for the review. Updated.

File: cmd/m4d/taxonomy.go (hunk 4 above)
Comment (roee88) [reply cmd/m4d/taxonomy.go:24]: Not that I mind too much in this case, but generally I disagree with the need to add these dummy docs to functions. 
Taking this one just as an example, this function is not exported, its name is descriptive, the pattern is known to most go programmers (this is how cobra works), and the logic itself of what the command does should actually be within the `Short`/`Long` arguments to the command. In any case, the code style requires function documentation to start with the function name.

**Comment by roee88** [comment]
@ronenkat  what is the blocker to this PR?

File: pkg/taxonomy/schema.go [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+		return err
+	}
+	path = filepath.Clean(path)
+
+	err = ValidateDraft4(path)
+	if err != nil {
+		return errors.Wrap(err, "Schema is not a valid DRAFT 4 JSON schema")
+	}
+
+	err = ValidateStructural(orig)
```
Comment (roee88) [inline pkg/taxonomy/schema.go:34]: Why orig? What happened when you used path?
Using a filepath without cleaning it first is considered a security risk, when we start using something like gosec it will capture those as errors (which reminds me to open an issue for it).

File: pkg/taxonomy/schema.go [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+	spec := fmt.Sprintf(`
+openapi: 3.0.1
+components:
+  schemas:
+    Object:
+      type: object
+      properties:
+        root:
+          $ref: '%s'
```
Comment (roee88) [inline pkg/taxonomy/schema.go:85]: This requires the file to be an object with a property for each root object to validate. We can take this into account in the base taxonomy without changes here. FYI @ritwikc let's discuss this separately.

File: pkg/taxonomy/schema_test.go [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+func TestValidateSchema_simple_success(t *testing.T) {
+	taxonomyFile := "../../test/taxonomy/catalog.structs.schema.json"
+	g := NewGomegaWithT(t)
+	err := ValidateSchema("./" + taxonomyFile)
+	g.Expect(err).NotTo(HaveOccurred())
+}
+
+// Validate a sample badly formed validation schema
+func TestValidateSchema_simple_fail(t *testing.T) {
+	taxonomyFile := "../../test/taxonomy/bad_catalog.structs.schema.json"
```
Comment (roee88) [inline pkg/taxonomy/schema_test.go:19]: Looking at those test files they seem to be quite large and it's not straight forward to see what is tested. Consider replacing with shorter files that are dedicated to testing. Not necessarily for this PR because we're changing things anyway.

File: pkg/taxonomy/schema_test.go [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+	g := NewGomegaWithT(t)
+	err := ValidateSchema("./" + taxonomyFile)
```
Comment (roee88) [inline pkg/taxonomy/schema_test.go:21]: Why the `./` prefix to an already relative path? Same for the TestValidateSchema_simple_success function.

File: cmd/taxonomy_verify.go [hunk 17]
```diff
@@ -0,0 +1,47 @@
+// Copyright 2020 IBM Corp.
+// SPDX-License-Identifier: Apache-2.0
```
[the remark this answers is not in the document]
Comment (ronenkat) [reply cmd/taxonomy_verify.go]: I think taxonomy_compile doesn't :- ) but happy to address.

File: cmd/taxonomy_verify.go [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+import (
+	"fmt"
+	"strings"
+
+	"github.com/mesh-for-data/mesh-for-data/pkg/taxonomy"
+	"github.com/spf13/cobra"
+	"helm.sh/helm/v3/cmd/helm/require"
+)
+
+// taxonomyCmd represents "schema" command option as part of the "taxonomy" commands
```
[the remark this answers is not in the document]
Comment (ronenkat) [reply cmd/taxonomy_verify.go]: sure. removing layers is good

File: cmd/taxonomy_verify.go [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+var schemaCmd = &cobra.Command{
+	Use:   "schema",
+	Short: "commands for working with taxonomy JSON schema files",
+	Args:  require.NoArgs,
+}
+
+// validateCmd  "validate" schema file command under the schema command
```
[the remark this answers is not in the document]
Comment (ronenkat) [reply cmd/taxonomy_verify.go]: Right. should have been taxonmy_validate.

File: pkg/taxonomy/schema_test.go (hunk 16 above)
Comment (ronenkat) [reply pkg/taxonomy/schema_test.go]: Right.

File: pkg/taxonomy/schema.go (hunk 13 above)
Comment (ronenkat) [reply pkg/taxonomy/schema.go]: +1

File: cmd/taxonomy_verify.go [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+			// which could be a path
+			return nil, cobra.ShellCompDirectiveDefault
+		}
+		// No more completions, so disable file completion
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	},
+	RunE: func(cmd *cobra.Command, args []string) error {
+		err := taxonomy.ValidateSchema(args[0])
+		if err == nil {
+			fmt.Println("Validate: " + strings.Join(args, " ") + " successfully")
```
[the remark this answers is not in the document]
Comment (ronenkat) [reply cmd/taxonomy_verify.go]: Thanks.

**Review by ronenkat — COMMENTED** [review]
Thank you for the feedback. Addressed.

**Review by roee88 — APPROVED** [review]
